Implemented a "group by trace" processor.

jpkrohling opened the pull request:
Signed-off-by: Juraci Paixão Kröhling <juraci@kroehling.de>

**Description:** This change adds a new 'groupbytrace' processor, as defined on #1309. This addresses the first item in the action plan.

**Link to tracking Issue:** #1309

**Testing:** extensive unit tests have been added

**Documentation:** sample configuration + readme added

jpkrohling commented on processor/groupbytraceprocessor/factory.go:
```diff
@@ -0,0 +1,107 @@
+		WaitDuration: defaultWaitDuration,
+
+		// not supported for now
+		DiscardOrphans: defaultDiscardOrphans,
```
I decided to keep these two out for this PR, as it was already big enough for a first proposal. Once this PR gets merged, I'll work on implementing these two features (discard orphans and store in disk).

nilebox replied:
I suggest to remove these options and add them in appropriate PRs with corresponding functionality.

jpkrohling replied:
I'll indeed remove them before this gets ready to be merged. Those options are here to show the direction this process is intended to go and to foster discussion :-)

jpkrohling commented:
@pjanotti, @bogdandrutu , @tigrannajaryan , would one of you please review this one? I marked it as draft for now, as there might be things to discuss, but from my perspective, this is ready for review.

jpkrohling commented:
@owais , would you be able to look into this one, please?

jrcamp commented on processor/groupbytraceprocessor/README.md:
```diff
@@ -0,0 +1,27 @@
+# Group by Trace processor
+
+Supported pipeline types: traces
+
```
Should specify it's in development.

jpkrohling replied:
Done!

jrcamp commented on processor/groupbytraceprocessor/config.go:
```diff
@@ -0,0 +1,46 @@
+
+	// NumTraces is the max number of traces to keep in memory waiting for the duration.
+	// Default: 1.000.000.
+	NumTraces int `mapstructure:"wait_duration"`
```
Looks like mapstructure tag is wrong?

jpkrohling replied:
To be honest, it's the first time I've seen `mapstructure` and wasn't able to quickly find documentation about it. It did seem a bit counter-intuitive to me at first, and as I checked other processors, I noticed that they use the same tag for literal types as well:

https://github.com/open-telemetry/opentelemetry-collector/blob/8693118652c9ccabef31ffc76320251cb0379fcf/processor/samplingprocessor/tailsamplingprocessor/config.go#L80-L95

https://github.com/open-telemetry/opentelemetry-collector/blob/8693118652c9ccabef31ffc76320251cb0379fcf/processor/batchprocessor/config.go#L23-L36

https://github.com/open-telemetry/opentelemetry-collector/blob/8693118652c9ccabef31ffc76320251cb0379fcf/processor/queuedprocessor/config.go#L23-L35

jrcamp replied:
I meant it's mapping to wait_duration instead of num_traces.

jpkrohling replied:
ouch, looks like I missed this twice :-) PR updated.

jpkrohling commented:
> Is this a common enough use case to go into core or should it go in contrib?

My idea is that once this processor gets stable enough, the tail-based sampler starts using it instead of having its own similar logic.

jpkrohling commented:
I just realized that I left a change to `service/defaultcomponents/defaults.go` out. Will amend this PR later to include this, but the change there is trivial and doesn't block the review.

nilebox commented on processor/groupbytraceprocessor/config.go:
```diff
@@ -0,0 +1,46 @@
+	configmodels.ProcessorSettings `mapstructure:",squash"`
+
+	// NumTraces is the max number of traces to keep in memory waiting for the duration.
+	// Default: 1.000.000.
```
nit: `1,000,000` format is more common in the English speaking world, I believe.
https://en.wikipedia.org/wiki/Decimal_separator#Examples_of_use

jpkrohling replied:
Sure, I can change, but is it ambiguous the way it is?

nilebox replied:
1_000_000 seems fine.

nilebox commented on processor/groupbytraceprocessor/config.go:
```diff
@@ -0,0 +1,46 @@
+	// Useful when the duration to wait for traces to complete is high.
+	// Default: false.
+	// Not yet implemented, and an error will be returned when this option is used.
+	StoreInDisk bool `mapstructure:"store_in_disk"`
```
Not a native speaker, but "store **on** disk" seems more correct.

jpkrohling replied:
"on disk", but "in memory". Go figure :-)

nilebox commented on processor/groupbytraceprocessor/config.go:
```diff
@@ -0,0 +1,46 @@
+	// StoreInDisk tells the processor to keep only the trace ID in memory, serializing the trace spans to disk.
+	// Useful when the duration to wait for traces to complete is high.
+	// Default: false.
+	// Not yet implemented, and an error will be returned when this option is used.
```
If it's not implemented, why do we need to add it right now? Better to add it in the PR adding support for storing on disk? The implementation of that doesn't seem trivial, so I would just remove this parameter for now.

jpkrohling replied:
I'll remove this before the final review.

nilebox commented on processor/groupbytraceprocessor/factory.go:
```diff
@@ -0,0 +1,107 @@
+
+	logger := params.Logger
+	if logger == nil {
+		logger = zap.NewNop()
```
Setting logger is responsibility of the code invoking `CreateTraceProcessor`.
Currently, logger is never nil: https://github.com/open-telemetry/opentelemetry-collector/blob/db926fd7f488bfabe0a4558578b2d59c48cc56a8/service/builder/pipelines_builder.go#L159

If you need this for unit tests, just init `ProcessorCreateParams` with nop logger in the test.

jpkrohling replied:
"Currently" being the keyword. I guess I'm just too used to do defensive programming, but I can surely remove it if necessary.

nilebox replied:
None of the existing processors check for that, so "currently" is an implicit contract anyway that would break all components if changed to nil, so I would just remove it.
Go doesn't support null safety, so I think it's fine to just panic with nil pointer dereference wherever the logger is actually used.

jpkrohling replied:
Removed.

nilebox commented on processor/groupbytraceprocessor/processor.go:
```diff
@@ -0,0 +1,344 @@
+			sp.processExpired(traceID)
+		case item := <-sp.received:
+			// in this implementation, this won't return errors
+			batch := pdata.TracesToOtlp(item)
```
Why do you need to convert to OTLP model rather than iterating through `item.ResourceSpans()`?

jpkrohling replied:
I'll double check, but I think I needed something from the protobuf models. It might have something to do with the trace ID.

nilebox commented on processor/groupbytraceprocessor/processor.go:
```diff
@@ -0,0 +1,344 @@
+func (sp *groupByTraceProcessor) onReleased() {
+	for item := range sp.released {
+		// what should we do with the context?
+		// should we have attempted to store the first context we saw? the last? none?
```
You can't use first, last or any other context from the individual call to `ConsumeTraces` because it may have been cancelled by the time you decide to asynchronously export data.

jpkrohling replied:
Yeah, and whoever set the context might have a different notion of context as we have here. Other places (batch processor?) just discards the context, and I think it might indeed be the right approach here as well.

nilebox commented on processor/groupbytraceprocessor/processor.go:
```diff
@@ -0,0 +1,344 @@
+
+// Shutdown is invoked during service shutdown.
+func (sp *groupByTraceProcessor) Shutdown(ctx context.Context) error {
+	sp.stopped <- ctx
```
Do you flush all buffered traces on Shutdown?

jpkrohling replied:
Most of the operations are executed as part of the main loop, so, the effect should be minimal in this implementation. When real on disk storage is implemented, the disk write might be async, so, a shutdown signal will need to be propagated down. After the shutdown call is received, the traces waiting for the duration will be discarded once their wait duration elapses.

nilebox replied:
> Most of the operations are executed as part of the main loop

You forward traces to the nextConsumer asynchronously via `go sp.markAsReleased(traceID)` after `wait_duration` is expired, which is not part of the main loop.

> After the shutdown call is received, the traces waiting for the duration will be discarded once their wait duration elapses.

The default `wait_duration` is 1 second and can be configured to be even bigger.

@tigrannajaryan @bogdandrutu what is expected graceful shutdown behavior? I would assume that we should flush all buffers to export any remaining data rather than just dropping them.

jpkrohling replied:
IMO, those traces should be discarded -- we have been asked explicitly by the user to wait for a specific duration before dispatching the trace. Sending it before will potentially dispatch incomplete traces, breaking one of the use cases for this processor.

jpkrohling replied:
> the trace may not complete within expected duration, so consumer must be tolerant to incomplete traces IMO.

I think we have different expectations here: when using this processor, the intention of our users is to receive complete traces, by waiting a specific amount of time before dispatching it. My expectation is that incomplete traces would be discarded, your expectation is that incomplete traces should be passed down the pipeline. Should we make a flag to control this? I can see that users will be confused either way if we don't document this well and provide a way to override the default behavior.

nilebox commented on processor/groupbytraceprocessor/processor.go:
```diff
@@ -0,0 +1,344 @@
+
+func (sp *groupByTraceProcessor) processBatch(batch []*v1.ResourceSpans) error {
+	for _, trace := range batch {
+		// TODO: should we collect all errors?
```
We should at least log them

nilebox commented on processor/groupbytraceprocessor/processor.go:
```diff
@@ -0,0 +1,344 @@
+		return errNilTraceID
+	}
+
+	hashedTraceID := hash(traceID)
```
If you keep the `pdata` model instead of converting to OTLP, you will be able to use https://github.com/open-telemetry/opentelemetry-collector/blob/db926fd7f488bfabe0a4558578b2d59c48cc56a8/consumer/pdata/trace.go#L107 and won't need hashing.

jpkrohling commented:
> Using pdata model without conversion to OTLP might be better, and that will allow you will to use TraceID.String() instead of hashing.

The last commit changes the PR to use the internal pdata model instead of OTLP. The performance gain seems to be consistent and relevant:

Before the last commit (f81445e701d508d33c6f8be79e509f6c7683e9ac), 3 benchmark runs:
```
$ go test -benchmem -run="^$" ./processor/groupbytraceprocessor -bench "^(BenchmarkConsumeTracesCompleteOnFirstBatch)$"
goos: linux
goarch: amd64
pkg: go.opentelemetry.io/collector/processor/groupbytraceprocessor
BenchmarkConsumeTracesCompleteOnFirstBatch-12    	 1000000	      7714 ns/op	   12468 B/op	     159 allocs/op
PASS
ok  	go.opentelemetry.io/collector/processor/groupbytraceprocessor	7.779s
$ go test -benchmem -run="^$" ./processor/groupbytraceprocessor -bench "^(BenchmarkConsumeTracesCompleteOnFirstBatch)$"
goos: linux
goarch: amd64
pkg: go.opentelemetry.io/collector/processor/groupbytraceprocessor
BenchmarkConsumeTracesCompleteOnFirstBatch-12    	 1000000	      8129 ns/op	   13159 B/op	     168 allocs/op
PASS
ok  	go.opentelemetry.io/collector/processor/groupbytraceprocessor	8.193s
$ go test -benchmem -run="^$" ./processor/groupbytraceprocessor -bench "^(BenchmarkConsumeTracesCompleteOnFirstBatch)$"
goos: linux
goarch: amd64
pkg: go.opentelemetry.io/collector/processor/groupbytraceprocessor
BenchmarkConsumeTracesCompleteOnFirstBatch-12    	 1000000	      8736 ns/op	   13946 B/op	     178 allocs/op
PASS
ok  	go.opentelemetry.io/collector/processor/groupbytraceprocessor	8.802s
```

After:
```
$ go test -benchmem -run="^$" ./processor/groupbytraceprocessor -bench "^(BenchmarkConsumeTracesCompleteOnFirstBatch)$"
goos: linux
goarch: amd64
pkg: go.opentelemetry.io/collector/processor/groupbytraceprocessor
BenchmarkConsumeTracesCompleteOnFirstBatch-12    	 1000000	      3377 ns/op	    6463 B/op	      54 allocs/op
PASS
ok  	go.opentelemetry.io/collector/processor/groupbytraceprocessor	3.438s
$ go test -benchmem -run="^$" ./processor/groupbytraceprocessor -bench "^(BenchmarkConsumeTracesCompleteOnFirstBatch)$"
goos: linux
goarch: amd64
pkg: go.opentelemetry.io/collector/processor/groupbytraceprocessor
BenchmarkConsumeTracesCompleteOnFirstBatch-12    	 1000000	      2781 ns/op	    5020 B/op	      44 allocs/op
PASS
ok  	go.opentelemetry.io/collector/processor/groupbytraceprocessor	2.842s
$ go test -benchmem -run="^$" ./processor/groupbytraceprocessor -bench "^(BenchmarkConsumeTracesCompleteOnFirstBatch)$"
goos: linux
goarch: amd64
pkg: go.opentelemetry.io/collector/processor/groupbytraceprocessor
BenchmarkConsumeTracesCompleteOnFirstBatch-12    	 1000000	      3560 ns/op	    7062 B/op	      59 allocs/op
PASS
ok  	go.opentelemetry.io/collector/processor/groupbytraceprocessor	3.636s
```

nilebox commented on processor/groupbytraceprocessor/processor.go:
```diff
@@ -0,0 +1,344 @@
+	// used for the ring buffer
+	index     int
+	ids       []pdata.TraceID
+	idToIndex map[string]int // key is hash(traceID), value is the index on the 'ids' slice
```
Please update the comment, it's not a hash anymore.

nilebox commented on processor/groupbytraceprocessor/hash.go:
```diff
@@ -0,0 +1,27 @@
+
+var seed = maphash.MakeSeed()
+
+func hash(traceID []byte) uint64 {
```
You shouldn't need this function (and file) anymore.

nilebox commented on processor/groupbytraceprocessor/hash_test.go:
```diff
@@ -0,0 +1,33 @@
+	"github.com/stretchr/testify/assert"
+)
+
+func TestHashStableOverInvocations(t *testing.T) {
```
Can be removed.

nilebox commented on processor/groupbytraceprocessor/processor.go:
```diff
@@ -0,0 +1,344 @@
+		return errNilTraceID
+	}
+
+	hashedTraceID := traceID.String()
```
```suggestion
	sTraceID := traceID.String()
```

nilebox commented on processor/groupbytraceprocessor/processor.go:
```diff
@@ -0,0 +1,344 @@
+	return nil
+}
+
+func (sp *groupByTraceProcessor) markAsReleased(traceID []byte) error {
```
```suggestion
func (sp *groupByTraceProcessor) markAsReleased(traceID pdata.TraceID) error {
```

nilebox commented on processor/groupbytraceprocessor/storage_memory.go:
```diff
@@ -0,0 +1,99 @@
+		return errNilResourceSpans
+	}
+
+	hashedID := traceID.String()
```
`sTraceID` here and in other places, you don't compute hash anymore.

nilebox replied:
Have you forgotten to push some changes? Still named `hashedID`.

jpkrohling replied:
Indeed, the storage was missing. This time I checked with `grep hash processor/groupbytraceprocessor -ir` and `hashID` isn't being used anymore.

jpkrohling commented:
@nilebox, @owais: other than the clarification around the shutdown behavior, is there anything pending for this PR?

nilebox commented on processor/groupbytraceprocessor/processor.go:
```diff
@@ -0,0 +1,343 @@
+	// see if the ring has an item already
+	idAtRingItem := sp.ids[sp.index]
+	if idAtRingItem != nil {
+		// delete from the map and erase its memory entry
```
Why not to do it together with the storage removal in processing items from the `removed` channel?
Optimization for potentially slow storage?

jpkrohling replied:
Yes, storage is potentially blocking (just not in the current implementation).

nilebox commented on processor/groupbytraceprocessor/processor.go:
```diff
@@ -0,0 +1,343 @@
+
+	// signal that the trace is ready to be released
+	sp.logger.Debug("trace marked as released", zap.Stringer("traceID", traceID))
+	sp.released <- trace
```
Shouldn't you also remove trace from the storage here?
```go
sp.removed <- traceID
```
This function gets called from `processExpired`, where TraceID is removed from `idToIndex` and `ids`, so we lost the pointer to item in the storage already, and this is our last chance to collect the garbage AFAICT.

jpkrohling replied:
The storage has a copy of the item, it isn't a reference to the same object.

nilebox replied:
Perhaps I wasn't clear. I didn't literally mean GC in Go.

My concern is there is no invocation of `sp.st.delete(traceID)` in this code flow.
The only way to trigger deletion of trace from storage at the moment is to push traceID to the `removed` channel, otherwise it will be stored "forever" (or far longer than needed).

The only place where you do `sp.removed <- idAtRingItem` is in the `processTrace` function, when the ring is overflown, which normally should never happen.

jpkrohling replied:
Looks like I missed your last comment while editing my previous one. You are right, and the code has been fixed :-)

nilebox replied:
Thanks for fixing this!
I still think that you need to introduce some abstractions to simplify the code and make it easier to catch this sort of errors. 

Please consider extracting the "ring buffer" into a separate struct and introducing an "event queue" with just one channel instead of 4.

nilebox commented on processor/groupbytraceprocessor/processor.go:
```diff
@@ -0,0 +1,343 @@
+	idAtRingItem := sp.ids[sp.index]
+	if idAtRingItem != nil {
+		// delete from the map and erase its memory entry
+		delete(sp.idToIndex, idAtRingItem.String())
```
You do this in 2 places, would prefer to consolidate the logic to make sure you have only one loop deleting items if possible. Same below, I think you need to at least extract any modifications in `idToIndex` and `ids` to separate functions, and same for reading from them, since they must be in sync at all times.

Probably worth implementing a separate struct for doing ring manipulations (get, add, delete)

jpkrohling replied:
> to make sure you have only one loop deleting items

Both operations are part of the same loop, just in different code branches (`processExpired` and `processTrace`). I think it's overkill to keep abstract away a simple map key deletion, it would bring unnecessary complexity.

nilebox replied:
It's probably easy for you to understand since you wrote this code, but for me having a function `deleteTracePointer(sTraceID)` is much easier to read than seeing
```go
	delete(sp.idToIndex, sTraceID)
	sp.ids[indexToRemove] = nil
```
```go
	delete(sp.idToIndex, idAtRingItem.String())
	sp.ids[sp.index] = nil
```
which look different but do 100% the same, and making sure they indeed do 100% the same is not great.

jpkrohling replied:
There was a small optimization when having the code in its original format (one `idToIndex` lookup on `processExpired`, now they are two), but in name of legibility, I followed your suggestion and made the clearing of traces from the slice and from the map part of a single function.

nilebox commented on processor/groupbytraceprocessor/processor.go:
```diff
@@ -0,0 +1,343 @@
+	st storage
+
+	// traces received from the previous processors
+	received chan pdata.Traces
```
Would it potentially be easier to read if you had just one channel with "event type" attached to event (received / expired / released / removed), and a single loop processing all of them?
It's hard to keep mental model of juggling with 4 channels.

I see that you are also using different data types for each channel, so not sure if that's feasible.

Alternatively, you may also consider making a "state machine" given the transition graph you have documented above.

jpkrohling replied:
I like the idea of the state machine. I'll give this a try.

jpkrohling replied:
I just added another commit to this PR adding an event machine. I think it's now more readable as the concerns are separated. The performance also seems to be a bit better:

```console
$ go test -benchmem -run=^$ go.opentelemetry.io/collector/processor/groupbytraceprocessor -bench "^(BenchmarkConsumeTracesCompleteOnFirstBatch)$"
goos: linux
goarch: amd64
pkg: go.opentelemetry.io/collector/processor/groupbytraceprocessor
BenchmarkConsumeTracesCompleteOnFirstBatch-12    	 1393412	       848 ns/op	    1107 B/op	      18 allocs/op
PASS
ok  	go.opentelemetry.io/collector/processor/groupbytraceprocessor	2.085s
```

nilebox commented on processor/groupbytraceprocessor/processor.go:
```diff
@@ -0,0 +1,343 @@
+		nextConsumer: nextConsumer,
+		config:       config,
+		st:           st,
+		ids:          make([]pdata.TraceID, config.NumTraces),
```
On a second thought, given that this is a ring buffer, there must be sufficient extra unused space to avoid data loss, i.e. this is by design.

jpkrohling commented on processor/groupbytraceprocessor/processor.go:
```diff
@@ -0,0 +1,343 @@
+
+// newGroupByTraceProcessor returns a new processor.
+func newGroupByTraceProcessor(logger *zap.Logger, st storage, nextConsumer consumer.TraceConsumer, config Config) (*groupByTraceProcessor, error) {
+	if nextConsumer == nil {
```
Interesting, I got this from the sampling processor:

https://github.com/open-telemetry/opentelemetry-collector/blob/f954863531b0949fc53f6ce9462a5285a184f7c8/processor/samplingprocessor/tailsamplingprocessor/processor.go#L76-L78

So, either this wasn't the case in the past, or this behavior isn't part of the contract and processors should protect themselves from next consumers being nil. Given that the error is actually in `componenterror`, I'd assume that this is indeed something that might happen. Perhaps it's protecting against downstream collector distributions (contrib)?

nilebox replied:
There is no point in having processor without exporter, pipeline is incomplete without it.

jpkrohling commented on processor/groupbytraceprocessor/processor.go:
```diff
@@ -0,0 +1,343 @@
+		sp.removed <- idAtRingItem
+
+		// TODO: do we want another channel that receives evicted items? record a metric perhaps?
+		sp.logger.Debug("trace evicted", zap.Stringer("traceID", idAtRingItem))
```
Should we care about throttling here? I'm afraid that, once this condition happens, quite a bunch of traces will be evicted until the collector was able to ingest enough of the in-queue traces.

jpkrohling replied:
I'm marking this as resolved, as this is `Info` now and zap does come with throttling in the default production setting. I didn't set the level to warn, as it apparently prints a stack trace, which might be too much.

jpkrohling commented on processor/groupbytraceprocessor/processor.go:
```diff
@@ -0,0 +1,343 @@
+	}
+
+	sp := &groupByTraceProcessor{
+		logger:       logger.With(zap.String("processor", "groupbytrace")),
```
You are absolutely right:

```
2020-07-30T08:51:01.186+0200	DEBUG	groupbytraceprocessor/processor.go:233	releasing trace	{"component_kind": "processor", "component_type": "groupbytrace", "component_name": "groupbytrace", "processor": "groupbytrace", "traceID": "00000000000000007e8c30f284eb3c98"}
```

nilebox commented on processor/groupbytraceprocessor/ring_buffer.go:
```diff
@@ -0,0 +1,69 @@
+
+import "go.opentelemetry.io/collector/consumer/pdata"
+
+type ringBuffer struct {
```
This is exactly what I suggested, looks good 👍

nilebox commented on processor/groupbytraceprocessor/storage_memory.go:
```diff
@@ -0,0 +1,101 @@
+	} else {
+		newTrace := pdata.NewResourceSpans()
+		newTrace.InitEmpty()
+		trace.CopyTo(newTrace)
```
Why do you make copies here but append without copying when `existing` is present?
Unless there is a specific reason, I think the whole logic can be unified as follows:
```go
dest, ok := st.content[sTraceID]
if !ok {
  dest := pdata.NewResourceSpans()
  dest.InitEmpty()
  st.content[sTraceID] = dest
}

// copy spans
trace.InstrumentationLibrarySpans().CopyTo(dest.InstrumentationLibrarySpans())
```

jpkrohling replied:
That's an oversight. It should be doing a copy of the spans. I just changed the code to do it and added a test.

nilebox commented on processor/groupbytraceprocessor/storage_memory.go:
```diff
@@ -0,0 +1,101 @@
+		return pdata.NewResourceSpans(), nil
+	}
+
+	result := pdata.NewResourceSpans()
```
Why do you need to make a copy? Preventing potential mutations?
Would be more efficient to just return `trace` and leave a comment that you don't expect any mutations to happen outside.

jpkrohling replied:
It's more efficient, but a future implementation (in disk storage) will have a different behavior and I prefer to have consistent behavior across implementations.

nilebox commented on processor/groupbytraceprocessor/storage_memory.go:
```diff
@@ -0,0 +1,101 @@
+
+	st.Lock()
+	trace := st.content[sTraceID]
+	result := pdata.NewResourceSpans()
```
Same here, creating a copy seems unnecessary

nilebox commented on processor/groupbytraceprocessor/processor.go:
```diff
@@ -0,0 +1,328 @@
+		return nil, errNoSpans
+	}
+
+	return spans.At(0).TraceID(), nil
```
Could spans in one slice have different traceIDs?

jpkrohling replied:
Up to this moment, I was under the impression that `ResourceSpans` would hold only spans related to the same trace, but I just realized that my assumption might be wrong. 

https://github.com/open-telemetry/opentelemetry-proto/blob/dfa8c325321d55cd9e27fe6e2c7bf927bb4d9ca3/opentelemetry/proto/trace/v1/trace.proto#L27-L56

nilebox replied:
I'm afraid you may need to group spans by traceID into separate batches and process each separately.

nilebox replied:
@bogdandrutu could you please confirm which assumption is correct?

nilebox commented on processor/groupbytraceprocessor/processor.go:
```diff
@@ -0,0 +1,328 @@
+		return errNilTraceID
+	}
+
+	if sp.ringBuffer.contains(traceID) {
```
You can rewrite this with a reverse condition to avoid code duplication:
```go
if !sp.ringBuffer.contains(traceID) {
  // Allocate new trace
  // 1: sp.ringBuffer.put(traceID), handle evicted
  // 2: time.AfterFunc(...)
}

// 3: sp.addSpans
```

The 1+2 steps could also be extracted into a separate function `createTraceIfMissing`, including the `ringBuffer.contains` check - then you won't need bother with ring buffer in this function at all.

nilebox replied:
This is a minor suggestion, the code looks readable.

nilebox commented on processor/groupbytraceprocessor/processor.go:
```diff
@@ -0,0 +1,328 @@
+
+	// this might block, but we don't need to wait
+	sp.logger.Debug("marking the trace as released", zap.Stringer("traceID", traceID))
+	go sp.markAsReleased(traceID)
```
Why do you need to start a goroutine here, given that you have allocated `expired` channel with a buffer of 50?
Do you expect the number of goroutines go higher than that?

jpkrohling replied:
`markAsReleased` calls `sp.st.get()`, which is potentially a blocking operation. As we don't actually have to wait for the outcome of `markAsReleased` to do further processing, we just run it in a separate routine.

jpkrohling commented:
Thank you @nilebox for the review rounds and for your patience. I believe I addressed all issues, except for the ones that deserve further discussion, namely:

- [ ] what should happen with in-flight traces when the processor is shutting down? Should we just discard them? Should we flush them immediately, blocking until they are consumed, possibly honoring the context's deadline? 
- [ ] should the processor protect itself against nil next consumers? Other processors seem to be doing it, others don't. The fact that there is an explicit error called `componenterror.ErrNilNextConsumer` makes me wonder what the contract really is, even though it doesn't make sense to have a processor without next consumer (which should at least be an exporter)
- [ ] can `ResourceSpans/InstrumentationLibrarySpans` contain spans from multiple traces?

nilebox commented on processor/groupbytraceprocessor/processor.go:
```diff
@@ -254,32 +217,25 @@ func (sp *groupByTraceProcessor) markAsReleased(traceID pdata.TraceID) error {
+		typ:     traceReleased,
+		payload: trace,
+	}, event{
+		typ:     traceRemoved,
```
`traceRemoved` is now the only event type emitted in 2 places.
It works fine, but I was wondering if we can group them differently.

The real events which trigger a change are `received` (in), `expired` and `evicted` (out). Others are parts of their processing:
- "released" is part of `onExpired`.
- "removed" is part of `onExpired` or `onEvicted`

Would there be a negative performance impact if we grouped them this way?
If you don't like this idea, feel free to leave it as it is.

nilebox replied:
Actually, "evicted" might also be part of `onReceived`, so there would be just 2 event types - `received` and `expired`, but that might be an overkill.

jpkrohling commented:
The last commit addresses the scenario where resource spans contain spans from multiple traces. There are only two points open: shutdown behavior and nil next consumer.

nilebox commented on processor/groupbytraceprocessor/processor.go:
```diff
@@ -116,69 +113,64 @@ func (sp *groupByTraceProcessor) Shutdown(ctx context.Context) error {
-		return fmt.Errorf("couldn't extract trace ID from trace: %w", err)
-	}
+	for _, batch := range splitByTrace(rs) {
+		traceID := batch.traceID
```
Please extract this entire block into a separate `processBatch` (or similar) function.
This will reduce indentation of the code and will improve its overall readability.

jpkrohling replied:
Done.

nilebox commented:
The code looks good to me now, but need to check the remaining open questions with maintainers of the repo.
Also not sure if this processor should live in the core repo or contrib.

jpkrohling commented:
>  Also not sure if this processor should live in the core repo or contrib.

This has been asked before by @jrcamp, and here's my answer:

> My idea is that once this processor gets stable enough, the tail-based sampler starts using it instead of having its own similar logic.

nilebox commented on processor/groupbytraceprocessor/processor.go:
```diff
@@ -128,50 +123,57 @@ func (sp *groupByTraceProcessor) processResourceSpans(rs pdata.ResourceSpans) er
-			// we are done with this trace, move on
-			return nil
+		if err := sp.processBatch(batch); err != nil {
+			sp.logger.Info("failed to process batch", zap.Error(err))
```
Might be useful to add traceID associated with the batch into message

jpkrohling replied:
Great idea. Added.

jpkrohling commented:
The last commit also removed the nil check on the next consumer, as I agree that it doesn't really make sense for a processor to not have a next consumer.

The only remaining point is about the shutdown then.

@tigrannajaryan, @bogdandrutu, would you please help us on making a decision here? [Basically](https://github.com/open-telemetry/opentelemetry-collector/pull/1362#issuecomment-666402837): 

>  what should happen with in-flight traces when the processor is shutting down? Should we just discard them? Should we flush them immediately, blocking until they are consumed, possibly honoring the context's deadline?

jpkrohling commented:
@nilebox would it be OK if I create an issue to discuss the shutdown behavior, implementing a fix in a follow-up PR?

nilebox commented:
@jpkrohling sure, I will approve this PR then once you create an issue. Still requires a maintainer to look at it and merge though :)

jpkrohling commented:
Created #1465

bogdandrutu commented on processor/groupbytraceprocessor/factory.go:
```diff
@@ -0,0 +1,100 @@
+type Factory struct {
+}
+
+var _ component.ProcessorFactory = (*Factory)(nil)
```
Please use the helper factory :), reduces the code and makes things consistent

jpkrohling replied:
Done!

jpkrohling commented:
I addressed @bogdandrutu's comment and squashed all the commits, since this was probably the last change in this PR.

jpkrohling commented:
@nilebox , @bogdandrutu is there anything extra to be done with this PR? If not, would you mind merging it?
